File: src/commands/schedule.ts
import { queryScheduler, startScheduler, stopScheduler } from "../platform/scheduler.ts";
import type { RuntimeOptions } from "../types.ts";

export async function runSchedule(
  action: "start" | "resume" | "stop" | "pause" | "status",
  options: RuntimeOptions = {},
): Promise<{ exitCode: number; stdout: string[]; stderr: string[] }> {
  const stdout: string[] = [];
  const stderr: string[] = [];

  try {
    const normalized = action === "resume" ? "start" : action === "pause" ? "stop" : action;
    const scheduler = normalized === "start"
      ? await startScheduler(options)
      : normalized === "stop"
      ? await stopScheduler(options)
      : await queryScheduler(options);

    if (normalized === "start") stdout.push("Scheduled backups started.");
    if (normalized === "stop") stdout.push("Scheduled backups stopped.");

    if (scheduler.labels.length === 0) {
      stdout.push("No scheduled jobs.");
    } else {
      for (const label of scheduler.labels) {
        const state = scheduler.states[label] ?? "not_installed";
        stdout.push(`${label}: ${state}`);
      }
    }

    return { exitCode: 0, stdout, stderr };
  } catch (error) {
    stderr.push(error instanceof Error ? error.message : String(error));
    return { exitCode: 1, stdout, stderr };
  }
}

File: rust/zen-backup/src/commands/schedule.rs
use crate::platform::{self, SchedulerState};

pub struct CommandOutput {
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
}

pub fn run_schedule(action: &str) -> CommandOutput {
    let normalized = match action {
        "resume" => "start",
        "pause" => "stop",
        other => other,
    };
    let mut stdout = Vec::new();

    let status = match normalized {
        "start" => platform::start_scheduler(),
        "stop" => platform::stop_scheduler(),
        "status" => platform::query_scheduler(),
        _ => {
            return CommandOutput {
                exit_code: 1,
                stdout: String::new(),
                stderr: "schedule action must be start|resume|stop|pause|status".to_string(),
            }
        }
    };

    if normalized == "start" {
        stdout.push("Scheduled backups started.".to_string());
    }
    if normalized == "stop" {
        stdout.push("Scheduled backups stopped.".to_string());
    }

    match status {
        Ok(scheduler) => {
            if scheduler.labels.is_empty() {
                stdout.push("No scheduled jobs.".to_string());
            } else {
                for (label, state) in scheduler.states {
                    let state_text = match state {
                        SchedulerState::Active => "active",
                        SchedulerState::Paused => "paused",
                        SchedulerState::NotInstalled => "not_installed",
                    };
                    stdout.push(format!("{label}: {state_text}"));
                }
            }
            CommandOutput {
                exit_code: 0,
                stdout: stdout.join("\n"),
                stderr: String::new(),
            }
        }
        Err(err) => CommandOutput {
            exit_code: 1,
            stdout: String::new(),
            stderr: err,
        },
    }
}
